bitpacker/src/blocked_bitpacker.rs

```diff
@@ -0,0 +1,176 @@
+use crate::BitUnpacker;
+
+use super::{bitpacker::BitPacker, compute_num_bits};
+
+const BLOCK_SIZE: usize = 128;
+
+/// `BlockedBitpacker` compresses data in blocks of
+/// 128 elements, while keeping an index on it
+///
+#[derive(Debug, Clone)]
+pub struct BlockedBitpacker {
+    // bitpacked blocks
+    compressed_blocks: Vec<u8>,
+    // uncompressed data, collected until BLOCK_SIZE
+    cache: Vec<u64>,
+    offset_and_bits: Vec<BlockedBitpackerEntryMetaData>,
+}
+
+/// `BlockedBitpackerEntryMetaData` encodes the
+/// offset and bit_width into a u64 bit field
+///
+/// This saves some space, since 7byte is more
+/// than enough and also keeps the access fast
+/// because of alignment
+#[derive(Debug, Clone, Default)]
+struct BlockedBitpackerEntryMetaData {
+    encoded: u64,
+    base_value: u64,
+}
+
+impl BlockedBitpackerEntryMetaData {
+    fn new(offset: u64, num_bits: u8, base_value: u64) -> Self {
+        let encoded = offset | (num_bits as u64) << (64 - 8);
+        Self {
+            encoded,
+            base_value,
+        }
+    }
+    fn offset(&self) -> u64 {
+        (self.encoded << 8) >> 8
+    }
+    fn num_bits(&self) -> u8 {
+        (self.encoded >> 56) as u8
+    }
+    fn base_value(&self) -> u64 {
+        self.base_value
+    }
+}
+
+#[test]
+fn metadata_test() {
+    let meta = BlockedBitpackerEntryMetaData::new(50000, 6, 40000);
+    assert_eq!(meta.offset(), 50000);
+    assert_eq!(meta.num_bits(), 6);
+}
+
+impl BlockedBitpacker {
+    pub fn new() -> Self {
+        let mut compressed_blocks = vec![];
+        compressed_blocks.resize(8, 0);
+        Self {
+            compressed_blocks,
+            cache: vec![],
```

Comment:
it is more a buffer than a cache isn't it?

Reply:
yes that's a better name